src/tp_evaluation_engine.hpp

```diff
@@ -0,0 +1,434 @@
+      const bool on_intialization = false);
+  // Grow space for storing temporary computation.
+  virtual void GrowSpareNodeData(const size_t new_node_spare_count);
+  virtual void GrowSpareEdgeData(const size_t new_edge_spare_count);
+
+  // Copy all edge data from its pre_edge_id to post_edge_id.
+  virtual void CopyEdgeData(const EdgeId src_edge_id, const EdgeId dest_edge_id);
+
+  // ** Populate PVs
+
+  // Populate rootward and leafward PVs.
+  void PopulatePVs();
+  // Populate P-PVs in a Rootward Pass of DAG.
+  void PopulateRootwardPVs();
+  // Populate R-PVs in a Leafward Pass of DAG.
+  void PopulateLeafwardPVs();
+
+  // ** Access
+
+  PSVEdgeHandler &GetPVs() { return parsimony_pvs_; }
+  const PSVEdgeHandler &GetPVs() const { return parsimony_pvs_; }
+  EigenMatrixXd &GetMatrix() { return parsimonies_; }
+  const EigenMatrixXd &GetMatrix() const { return parsimonies_; }
+
+ protected:
+  // ** Scoring Helpers
+
+  // Compute the rootward P-PVs for given node.
+  void PopulateRootwardParsimonyPVForNode(const NodeId node_id);
+  // Compute the leafward R-PVs for given node.
+  void PopulateLeafwardParsimonyPVForNode(const NodeId node_id);
+  // Set the P-PVs to match the observed site patterns at the leaves.
+  void PopulateLeafParsimonyPVsWithSitePatterns();
+  // Calculate the PV for a given parent-child pair.
+  EigenVectorXd ParentPartial(EigenVectorXd child_partials);
+  // Sum P-PVs for right and left children of node 'node_id'
+  // In this case, we get the full P-PVs of the given node after all P-PVs
+  // have been concatenated into one SankoffPartialVector.
+  EigenVectorXd TotalPPartial(const EdgeId edge_id, const size_t site_idx);
+  EigenVectorXd TotalPPartial(const PVId edge_pleft_pvid, const PVId edge_pright_pvid,
+                              const size_t site_idx);
+  // Populate rootward P-PVs for given edge.
+  // Updates parent's Pleft PV with sum of left child P PVs and parent's PRight PV with
+  // sum of right child P PVs.
+  void PopulateRootwardParsimonyPVForEdge(const EdgeId parent_id,
+                                          const EdgeId left_child_id,
+                                          const EdgeId right_child_id);
+  // Populate leafward Q-PVs for given edge.
+  // Updates parent's Q PV by combining with sister's sum of P PVs.
+  void PopulateLeafwardParsimonyPVForEdge(const EdgeId parent_id,
+                                          const EdgeId left_child_id,
+                                          const EdgeId right_child_id);
+  // Calculates parsimony score on given edge across all sites.
+  // Takes the minimum element after taking sum of edge's P and Q PVs.
+  double ParsimonyScore(const EdgeId edge_id);
+  double ParsimonyScore(const PVId edge_q_pvid, const PVId edge_pleft_pvid,
+                        const PVId edge_pright_pvid);
+
+ protected:
+  // Partial Vector for storing Parsimony scores.
+  PSVEdgeHandler parsimony_pvs_;
+  // Tree parsimonies across all sites.
```

Comment:
Can you provide a little more detail? I don't follow what this is about.

Reply:
Sorry, this was temp value, that was filled by Cost Matrix.  Removed.